cmd/vicadmin/server.go

```diff
@@ -114,39 +191,151 @@ func (s *server) listenPort() int {
+			websession, _ := store.Get(r, sessionCookieKey)
+			timeNow, _ := usersess.created.MarshalText()
+			websession.Values["created"] = string(timeNow)
+			websession.Values["username"] = "root"
+			websession.Save(r, w)
+			handler(w, r)
+			return
+		}
+
+		c := websession.Values["created"]
+		if c == nil { // no cookie, so redirect to login
+			http.Redirect(w, r, "/authentication", 302)
+			return
+		}
+
+		// parse the cookie creation time
+		created, _ := time.Parse(time.RFC3339, c.(string))
+
+		if time.Since(created) > sessionExpiration { // cookie exists but is expired
+			http.Redirect(w, r, "/authentication?expired", 302)
+			return
+		}
+
+		// if the date on the cookie was valid, then the user is authenticated
+		handler(w, r)
+	}
+	s.mux.Handle(link, gorillacontext.ClearHandler(http.HandlerFunc(authHandler)))
+}
+
+// renders the page for login and handles authorization requests
+func (s *server) loginPage(res http.ResponseWriter, req *http.Request) {
+	defer trace.End(trace.Begin(""))
+	ctx := context.Background()
+	if req.Method == "POST" {
+		// take the form data and use it to try to authenticate with vsphere
+
+		// create a userconfig
+		userconfig := session.Config{
+			Insecure:       false,
+			Thumbprint:     rootConfig.Thumbprint,
+			Keepalive:      rootConfig.Keepalive,
+			ClusterPath:    rootConfig.ClusterPath,
+			DatacenterPath: rootConfig.DatacenterPath,
+			DatastorePath:  rootConfig.DatastorePath,
+			HostPath:       rootConfig.Config.HostPath,
+			PoolPath:       rootConfig.PoolPath,
+		}
+		user := url.UserPassword(req.FormValue("username"), req.FormValue("password"))
+		serviceURL, err := soap.ParseURL(rootConfig.Service)
+		if err != nil {
+			log.Errorf("vSphere service URL was not a valid format; parsing returned error: %s", err)
+		}
+		serviceURL.User = user
+		userconfig.Service = serviceURL.String()
+
+		// check login
+		usersession, err := vSphereSessionGet(&userconfig)
+		if err != nil {
+			// something went wrong or we could not authenticate
+			http.Redirect(res, req, "/authentication?unauthorized", 302)
+		}
+
+		// successful login above; user is authenticated
+		// log out
+		usersession.Client.Logout(context.Background())
+
+		// save user config locally
+		usersess := s.uss.Add(req.FormValue("username"), &userconfig)
+
+		// create a token to save as an encrypted & signed cookie
+		websession, _ := store.Get(req, sessionCookieKey)
+		timeNow, _ := usersess.created.MarshalText()
+		websession.Values["created"] = string(timeNow)
+		websession.Values["username"] = req.FormValue("username")
+		websession.Save(req, res)
+
+		// redirect to dashboard
+		http.Redirect(res, req, "/", 302)
+	}
+
+	// Render login page (shows up on non-POST requests):
+	sess, err := client(&rootConfig)
+	v := vicadmin.NewValidator(ctx, &vchConfig, sess)
+	tmpl, err := template.ParseFiles("auth.html")
+	err = tmpl.ExecuteTemplate(res, "auth.html", v)
+	if err != nil {
+		log.Errorf("Error parsing template: %s", err)
```

Comment:
redirects are normally 3xx; why don't we just return a response here, and not redirect?